rectools/models/nn/transformers/bert4rec.py

```diff
@@ -0,0 +1,363 @@
+#  Copyright 2025 MTS (Mobile Telesystems)
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import typing as tp
+from collections.abc import Hashable
+from typing import Dict, List, Tuple
+
+import numpy as np
+import torch
+
+from ..item_net import (
+    CatFeaturesItemNet,
+    IdEmbeddingsItemNet,
+    ItemNetBase,
+    ItemNetConstructorBase,
+    SumOfEmbeddingsConstructor,
+)
+from .base import (
+    TrainerCallable,
+    TransformerDataPreparatorType,
+    TransformerLightningModule,
+    TransformerLightningModuleBase,
+    TransformerModelBase,
+    TransformerModelConfig,
+    ValMaskCallable,
+)
+from .constants import MASKING_VALUE, PADDING_VALUE
+from .data_preparator import TransformerDataPreparatorBase
+from .net_blocks import (
+    LearnableInversePositionalEncoding,
+    PositionalEncodingBase,
+    PreLNTransformerLayers,
+    TransformerLayersBase,
+)
+
+
+class BERT4RecDataPreparator(TransformerDataPreparatorBase):
+    """Data Preparator for BERT4RecModel."""
+
+    train_session_max_len_addition: int = 0
+
+    item_extra_tokens: tp.Sequence[Hashable] = (PADDING_VALUE, MASKING_VALUE)
+
+    def __init__(
+        self,
+        session_max_len: int,
+        n_negatives: tp.Optional[int],
+        batch_size: int,
+        dataloader_num_workers: int,
+        train_min_user_interactions: int,
+        mask_prob: float,
+        shuffle_train: bool = True,
+        get_val_mask_func: tp.Optional[ValMaskCallable] = None,
+    ) -> None:
+        super().__init__(
+            session_max_len=session_max_len,
+            n_negatives=n_negatives,
+            batch_size=batch_size,
+            dataloader_num_workers=dataloader_num_workers,
+            train_min_user_interactions=train_min_user_interactions,
+            shuffle_train=shuffle_train,
+            get_val_mask_func=get_val_mask_func,
+        )
+        self.mask_prob = mask_prob
+
+    def _mask_session(self, ses: List[int]) -> Tuple[List[int], List[int]]:
+        masked_session = ses.copy()
+        target = ses.copy()
+        random_probs = np.random.rand(len(ses))
+        for j in range(len(ses)):
+            if random_probs[j] < self.mask_prob:
+                random_probs[j] /= self.mask_prob
+                if random_probs[j] < 0.8:
```

Comment:
You know hardcoded values are super dangerous and inconvenient. Let's please move them to params

Reply:
Fixed